### ISSUE TITLE
Conflicting types

### ISSUE THREAD

**Issue by antoscha** [hfp/libxsmm#331]
Hi!
I'm getting the following error during compilation with `make STATIC=0 MKL=0 BLAS=1`:
https://pastebin.com/NPpwZCDk
It doesn't appear in 1.10. My system is Arch Linux and up to date.

**Comment by hfp** [comment]
... to keep this [report](https://github.com/hfp/libxsmm/files/3131575/report.pdf) properly archived.

**Comment by hfp** [comment]
I am getting up Arch Linux. Thank you for the report!

**Comment by hfp** [comment]
First of all `make STATIC=0 MKL=0 BLAS=1` makes sense in an environment where MKL and another BLAS are installed (e.g., RefBLAS and/or OpenBLAS). Apparently, you have OpenBLAS installed since it is detected by the build-system and some definitions are added accordingly (`-D__BLAS=1 -D__CBLAS -D__OPENBLAS`). You also seem to have MKL installed, but you want to pick OpenBLAS. Btw, you could prefer RefBLAS over OpenBLAS using `OPENBLAS=0`.

On my plan but up-to-date Arch Linux (plus `git`, `make`, `gcc`, `gcc-fortran`, `python`, and `openblas`), I can reproduce the problem. I will keep you posted.

**Comment by hfp** [comment]
A harmless workaround is to run `make STATIC=0 MKL=0 BLAS=1 ECFLAGS=-DLIBXSMM_GEMM_CONST`. I will fix this properly and likely come up with an updated release.

**Comment by antoscha** [comment]
Yes, you're quite right, I have two BLAS implementation installed: MKL (via environmental module) and OpenBLAS. Due to the fact that I use MKL for very specific purposes only, I've setting `MKL=0` when installing LIBXSMM (it ignores environment and tries to link to MKL).

Thank you for the workaround! You're doing a great job!

**Comment by hfp** [comment]
I actually wonder why you at all link with BLAS when building LIBXSMM. Is this to ease linking your application and omitting BLAS there? Normally, under Linux there is no need to link LIBXSMM with any BLAS. You can decide later when linking your application. In fact, that's what you can see at the end of your build:

```text
--------------------------------------------------------------------------------
LIBXSMM is link-time agnostic with respect to a BLAS library!
Forcing a specific library can take away a user's choice.
--------------------------------------------------------------------------------
```

Perhaps an even better workaround is to use `make STATIC=0`.

**Comment by hfp** [comment]
Btw, the issue is fixed in the master revision.

**Comment by antoscha** [comment]
I'm doing this because of the cp2k build error that appears when BLAS is not linked to LIBXSMM:
`/usr/bin/ld: /usr/lib/libxsmm.so: undefined reference to 'dgemm_'
/usr/bin/ld: /usr/lib/libxsmm.so: undefined reference to 'sgemm_'`

**Comment by hfp** [comment]
I certainly never build CP2K as a dynamically linked executable. I will perhaps test this better. Anyhow, that's not the topic here.

> I'm doing this because of the cp2k build error that appears when BLAS is not linked to LIBXSMM

The envisioned solution was/is that you specify BLAS with CP2K's link line. Of course, you can leverage LIBXSMM which is linked to BLAS "to save this effort". However, since CP2K needs BLAS independent of libxsmm, I think it is the better solution to link BLAS when CP2K is linked. If you link BLAS with CP2K, it's somewhat dangerous since you could specify potentially conflicting BLAS libraries. I wonder if the problem is related to the library-order (aka BLAS-last). Perhaps you can attached the ARCH-file?

**Comment by antoscha** [comment]
Here it is:
[Linux-x86-64-arch.psmp.txt](https://github.com/hfp/libxsmm/files/3137883/Linux-x86-64-arch.psmp.txt)

**Comment by hfp** [comment]
Thank you for the ARCH-file. It's a good starting point for my own tests!

The proposal is that `-lblas` put after `-lxsmmf -lxsmm` would solve the issue in the first place i.e., you could omit to link BLAS into LIBXSMM.

**Comment by antoscha** [comment]
Thank you, it work now!

**Comment by hfp** [comment]
The issue is fixed in [v1.12](https://github.com/hfp/libxsmm/releases/tag/1.12).
